### PR TITLE
Synchronous compile

### PR DESCRIPTION
**Pull request by tbosch**

### CODE REVIEW

File: modules/@angular/compiler/src/compile_metadata.ts [hunk 1]
```diff
@@ -629,6 +656,7 @@ export class CompileTemplateMetadata {
     this.templateUrl = templateUrl;
     this.styles = isPresent(styles) ? styles : [];
     this.styleUrls = isPresent(styleUrls) ? styleUrls : [];
+    this.externalStylesheets = isPresent(externalStylesheets) ? externalStylesheets : [];
```
Comment (tbosch) [inline modules/@angular/compiler/src/compile_metadata.ts]: Use `normalizeArray`

File: modules/@angular/compiler/src/directive_normalizer.ts [hunk 2]
```diff
@@ -12,65 +12,89 @@ import {PromiseWrapper} from '../src/facade/async';
...
-                lifecycleHooks: directive.lifecycleHooks,
-                providers: directive.providers,
-                viewProviders: directive.viewProviders,
-                queries: directive.queries,
-                viewQueries: directive.viewQueries,
-                precompile: directive.precompile,
-                template: normalizedTemplate
-              }));
+    if (isPresent(directive.template.template)) {
+      let normalizedTemplate = this.normalizeLoadedTemplate(
```
Comment (tbosch) [inline modules/@angular/compiler/src/directive_normalizer.ts]: Add comment: `synchronous path`, or create a separate method

File: modules/@angular/compiler/src/directive_normalizer.ts [hunk 3]
```diff
@@ -12,65 +12,89 @@ import {PromiseWrapper} from '../src/facade/async';
...
+                (normalizedTemplate) =>
+                    _cloneDirectiveWithTemplate(directive, normalizedTemplate)));
   }
 
-  normalizeTemplate(directiveType: CompileTypeMetadata, template: CompileTemplateMetadata):
+  normalizeTemplateAsync(directiveType: CompileTypeMetadata, template: CompileTemplateMetadata):
       Promise<CompileTemplateMetadata> {
+    let templateContent: Promise<string>;
+    let templateUrl: string;
     if (isPresent(template.template)) {
```
Comment (tbosch) [inline modules/@angular/compiler/src/directive_normalizer.ts]: Never come here when we have inline templates

File: modules/@angular/compiler/src/directive_normalizer.ts [hunk 4]
```diff
@@ -12,65 +12,89 @@ import {PromiseWrapper} from '../src/facade/async';
...
-          directiveType, template, template.template, directiveType.moduleUrl));
+      templateContent = Promise.resolve(template.template);
+      templateUrl = directiveType.moduleUrl;
     } else if (isPresent(template.templateUrl)) {
-      var sourceAbsUrl = this._urlResolver.resolve(directiveType.moduleUrl, template.templateUrl);
-      return this._xhr.get(sourceAbsUrl)
-          .then(
-              templateContent => this.normalizeLoadedTemplate(
-                  directiveType, template, templateContent, sourceAbsUrl));
+      let sourceAbsUrl = this._urlResolver.resolve(directiveType.moduleUrl, template.templateUrl);
```
Comment (tbosch) [inline modules/@angular/compiler/src/directive_normalizer.ts]: inlikne `sourceAbsUrl` variable

File: modules/@angular/compiler/src/directive_normalizer.ts [hunk 5]
```diff
@@ -81,42 +105,86 @@ export class DirectiveNormalizer {
...
-            .map(url => this._urlResolver.resolve(templateAbsUrl, url))
-            .concat(templateMeta.styleUrls.filter(isStyleUrlResolvable)
-                        .map(url => this._urlResolver.resolve(directiveType.moduleUrl, url)));
-
-    var allResolvedStyles = allStyles.map(style => {
-      var styleWithImports = extractStyleUrls(this._urlResolver, templateAbsUrl, style);
-      styleWithImports.styleUrls.forEach(styleUrl => allStyleAbsUrls.push(styleUrl));
-      return styleWithImports.style;
-    });
+    var templateMetadataStyles = this.normalizeStylesheet(new CompileStylesheetMetadata({
```
Comment (tbosch) [inline modules/@angular/compiler/src/directive_normalizer.ts]: pass the tempalte directly

Comment (tbosch) [inline modules/@angular/compiler/src/directive_normalizer.ts]: Make `CompileStylesheetMetadata.moduleUrl` return the `type.moduleUrl`

File: modules/@angular/compiler/src/directive_normalizer.ts [hunk 6]
```diff
@@ -81,42 +105,86 @@ export class DirectiveNormalizer {
       var errorString = rootNodesAndErrors.errors.join('\n');
       throw new BaseException(`Template parse errors:\n${errorString}`);
     }
-
```
Comment (tbosch) [inline modules/@angular/compiler/src/directive_normalizer.ts]: Call visitor further down, when we normalize the styles

File: modules/@angular/compiler/src/directive_normalizer.ts [hunk 7]
```diff
@@ -81,42 +105,86 @@ export class DirectiveNormalizer {
...
+        .then((_) => Array.from(loadedStylesheets.values()));
+  }
+
+  normalizeStylesheet(stylesheet: CompileStylesheetMetadata): CompileStylesheetMetadata {
+    var allStyleUrls = stylesheet.styleUrls.filter(isStyleUrlResolvable)
+                           .map(url => this._urlResolver.resolve(stylesheet.moduleUrl, url));
+
+    var allStyles = stylesheet.styles.map(style => {
+      var styleWithImports = extractStyleUrls(this._urlResolver, stylesheet.moduleUrl, style);
+      styleWithImports.styleUrls.forEach(styleUrl => allStyleUrls.push(styleUrl));
```
Comment (tbosch) [inline modules/@angular/compiler/src/directive_normalizer.ts]: Use `...` to add without the `forEach`

File: modules/@angular/compiler/src/offline_compiler.ts [hunk 8]
```diff
@@ -44,29 +44,38 @@ export class OfflineCompiler {
...
       throw new BaseException('No components given');
     }
     var statements: o.DeclareVarStmt[] = [];
     var exportedVars: string[] = [];
     var moduleUrl = _ngfactoryModuleUrl(components[0].component.type);
+    var outputSourceModules: SourceModule[] = [];
     components.forEach(componentWithDirs => {
       var compMeta = <CompileDirectiveMetadata>componentWithDirs.component;
       _assertComponent(compMeta);
+      var fileSuffix = _splitSuffix(compMeta.type.moduleUrl)[1];
```
Comment (tbosch) [inline modules/@angular/compiler/src/offline_compiler.ts]: Rename `_splitSuffix` into `_splitLastSuffix`

File: modules/@angular/compiler/src/style_compiler.ts [hunk 9]
```diff
@@ -40,35 +48,39 @@ export class StyleCompiler {
...
-    var styleWithImports = extractStyleUrls(this._urlResolver, stylesheetUrl, cssText);
-    return this._compileStyles(
-        getStylesVarName(null), [styleWithImports.style], styleWithImports.styleUrls, isShimmed);
+    var externalStylesheets: CompiledStylesheet[] = [];
+    var componentStylesheet: CompiledStylesheet = this._compileStyles(comp, comp.template, true);
+    comp.template.externalStylesheets.forEach((stylesheetMeta) => {
+      var compiledStylesheet = this._compileStyles(comp, stylesheetMeta, false);
+      externalStylesheets.push(compiledStylesheet);
+    });
+    if (!componentStylesheet) {
```
Comment (tbosch) [inline modules/@angular/compiler/src/style_compiler.ts]: No need to throw

**Comment by vicb** [comment]
Nice PR ! LGTM

File: modules/@angular/compiler/src/compile_metadata.ts (hunk 1 above)
Comment (tbosch) [inline modules/@angular/compiler/src/compile_metadata.ts]: done

File: modules/@angular/compiler/src/directive_normalizer.ts (hunk 2 above)
Comment (tbosch) [inline modules/@angular/compiler/src/directive_normalizer.ts]: done

File: modules/@angular/compiler/src/directive_normalizer.ts (hunk 3 above)
Comment (tbosch) [inline modules/@angular/compiler/src/directive_normalizer.ts]: done

File: modules/@angular/compiler/src/directive_normalizer.ts (hunk 4 above)
Comment (tbosch) [inline modules/@angular/compiler/src/directive_normalizer.ts]: done

File: modules/@angular/compiler/src/directive_normalizer.ts (hunk 5 above)
Comment (tbosch) [inline modules/@angular/compiler/src/directive_normalizer.ts]: done

File: modules/@angular/compiler/src/directive_normalizer.ts (hunk 6 above)
Comment (tbosch) [inline modules/@angular/compiler/src/directive_normalizer.ts]: done

File: modules/@angular/compiler/src/directive_normalizer.ts (hunk 7 above)
Comment (tbosch) [inline modules/@angular/compiler/src/directive_normalizer.ts]: done

File: modules/@angular/compiler/src/offline_compiler.ts (hunk 8 above)
Comment (tbosch) [inline modules/@angular/compiler/src/offline_compiler.ts]: Done

**Comment by marcalj** [comment]
@tbosch I love you! You fixed couple of bugs with ngUpgrade migration!

1. Components with forwardRef make app don't bootstrap
2. Doesn't trigger Angular 1 initial router event, so the app doesn't "bootstrap" any component!!!

Thanks a lot! Cheers!

**Pull request merged**